<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исследование-надежности-заемщиков." data-toc-modified-id="Исследование-надежности-заемщиков.-1">Исследование надежности заемщиков.</a></span><ul class="toc-item"><li><span><a href="#Шаг-1.-Обзор-данных" data-toc-modified-id="Шаг-1.-Обзор-данных-1.1">Шаг 1. Обзор данных</a></span></li><li><span><a href="#Шаг-2.1-Заполнение-пропусков" data-toc-modified-id="Шаг-2.1-Заполнение-пропусков-1.2">Шаг 2.1 Заполнение пропусков</a></span></li><li><span><a href="#Шаг-2.2-Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Шаг-2.2-Проверка-данных-на-аномалии-и-исправления.-1.3">Шаг 2.2 Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Шаг-2.3.-Изменение-типов-данных." data-toc-modified-id="Шаг-2.3.-Изменение-типов-данных.-1.4">Шаг 2.3. Изменение типов данных.</a></span></li><li><span><a href="#Шаг-2.4.-Удаление-дубликатов." data-toc-modified-id="Шаг-2.4.-Удаление-дубликатов.-1.5">Шаг 2.4. Удаление дубликатов.</a></span></li><li><span><a href="#Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-1.6">Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Шаг-2.6.-Категоризация-дохода." data-toc-modified-id="Шаг-2.6.-Категоризация-дохода.-1.7">Шаг 2.6. Категоризация дохода.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-1.8">Ответы на вопросы.</a></span></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-1.9">Общий вывод:</a></span></li></ul></li></ul></div>

# Исследование надежности заемщиков.

**Цель исследования**

Определить, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. 
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Для этого, на основании статистики о платёжеспособности клиентов, переданной банком, необходимо установить:
1. Наличие зависимости между количеством детей и возвратом кредита в срок.
2. Наличие зависимости между семейным положением и возвратом кредита в срок.
3. Наличие зависимости между уровнем дохода и возвратом кредита в срок.
4. Влияние цели кредита на его возврат в срок.

**Ход исследования**
1. Обзор данных, переданных банком.
2. Предобработка данных.
3. Проверка гипотез.

## Шаг 1. Обзор данных

Составим первое представление о данных банка

In [130]:
# импорт библиотеки pandas
import pandas as pd

In [131]:
# чтение файла с данными и сохранение в df для локальной и сетевой работы
try:
    df = pd.read_csv('data.csv')
except FileNotFoundError:
    df = pd.read_csv('/datasets/data.csv')

In [132]:
# получение первых 10 строк таблицы df для обзора
display(df.head(10))

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [133]:
# получение общей информации о данных в таблице df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Согласно документации к данным, названия колонок отражают следующую информацию:

* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

На основании обзора данных установлено:

1. Всего в таблице 25525 сторок и 12 колонок.
2. Названия колонок соответствуют стилю.
3. Присутствуют данные типов: int, float, object. Значения в столбцах `days_employed` и `total_income` стоит заменить на int.
4. В столбцах `days_employed` и `total_income` встречаются пропуски, причем количество пропусков в обих столбцах одинаково.
5. В столбце `education` присутсвуют неявные дубликаты значений (одни и те же значения в разном регистре).
6. В столбце `days_employed` в большом количестве присутствуют отрицательные значения рабочего стажа. Положительные значения выглядят нереалистично (932 года стажа).
7. В одной таблице хранятся идентификаторы и расшифровки для уровня образования и семейного статуса.


**Выводы**

В каждой строке таблицы — данные о потенцальном заемщике. Часть колонок содержит личную информацию - семейное положение, образоавние, количество детей, трудовой стаж, возраст, пол. Другая чатсть имеет прямое отношение к кредиту и платежеспособности - цель кредита, тип занятости, уровень дохода, факты задолжностей.

Предварительно можно утверждать, что, данных достаточно для проверки гипотез. Но в одной из колонок встречаются отрицательные значения неясного происхождения, при этом положительные значения также выглядят нереалистично. В двух столбцах есть пропуски в данных. Два столбца с идентификаторами семейного положения и образования дублируют информацию столбцов с расшифровкой. Расшифровки стоит вынести в отдельные таблицы, оставив только идентификаторы.

Чтобы двигаться дальше, нужно устранить проблемы в данных.

## Шаг 2.1 Заполнение пропусков

Ещё раз посмотрим на пропуски в данных о трудовом стаже и доходе.

In [134]:
# подсчёт пропусков
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [135]:
# Проверим зависимость между пропусками
df[df['days_employed'].isna()].head(10)

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

Доля пропусков от общего количества данных для дохода и стажа будет одинакова. Найдем ее.

In [136]:
# подсчёт доли пропусков от общего числа данных
part_na = df['days_employed'].isna().mean()
print(f'Доля пропусков в данных о трудовом стаже и в данных о доходе составляет: {part_na:.2%}')

Доля пропусков в данных о трудовом стаже и в данных о доходе составляет: 10.10%


Наличие пропусков в данных о доходах напрямую связано с пропусками в данных о трудовом стаже. Вероятно, данные о трудовом стаже использовались для расчета информации о доходах.

Данные о даходах необходимы нам для дальнейшего анализа. 
Для начала проверим, есть ли зависимось среднего и медиального уровня дохода от категории занятости.

In [137]:
# смотрим распределение количества потенциальных заемщиков по типу занятости
display(df['income_type'].value_counts())

# смотрим среднее значение дохода для каждого типа занятости. Пропуски можно не удалять, т.к. метод mean() игнорирует их
display(df.groupby('income_type')['total_income'].mean())

# смотрим медиальное значение дохода для каждого типа занятости. Пропуски можно не удалять, т.к. метод median() игнорирует их
df.groupby('income_type')['total_income'].median()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
в декрете              1
студент                1
Name: income_type, dtype: int64

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        170898.309923
компаньон          202417.461462
пенсионер          137127.465690
предприниматель    499163.144947
сотрудник          161380.260488
студент             98201.625314
Name: total_income, dtype: float64

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64

Средний и медиальный уровень дохода для разных типов звнятости сильно разнится. Чтобы получить минимальное искажение данных о даходах, заполним пропуски медиальными значениями по соответсвующей категории занятости.

In [138]:
# создаем список видов занятости
income_types = list(df['income_type'].unique())

# для каждого вида занятости в списке считаем медиальное значение и подставляем вместо пропусков
for income_type in income_types:
    means_to_replace = df[df['income_type'] == income_type]['total_income']
    median_total_income = means_to_replace.median()
    df.loc[df['income_type'] == income_type, 'total_income'] = means_to_replace.fillna(median_total_income)

In [139]:
# проверяем число пропусков
print(df.isna().sum())

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64


In [140]:
# сверяем значение с медиальным на примере строк 12 и 26 для пенсионера (118514.486412) и госслужащего (150447.935283)
print(df.loc[12])
print(df.loc[26])

children                           0
days_employed                    NaN
dob_years                         65
education                    среднее
education_id                       1
family_status       гражданский брак
family_status_id                   1
gender                             M
income_type                пенсионер
debt                               0
total_income           118514.486412
purpose              сыграть свадьбу
Name: 12, dtype: object
children                          0
days_employed                   NaN
dob_years                        41
education                   среднее
education_id                      1
family_status       женат / замужем
family_status_id                  0
gender                            M
income_type             госслужащий
debt                              0
total_income          150447.935283
purpose                 образование
Name: 26, dtype: object


Пропуски в столбце `total_income` успешно заменены на медиальные значения по типу занятости `income_type` и могут быть использованы для дальнейших расчетов.

## Шаг 2.2 Проверка данных на аномалии и исправления.

В столбце `days_employed` в большом количестве присутствуют отрицательные значения рабочего стажа. При этом положительные значения выглядят нереалистично.

Чтобы оценить адекватность данных, переведем значения в годы.

In [141]:
df['days_employed_years'] = df['days_employed'] / 365
df.loc[:, ['days_employed', 'days_employed_years', 'dob_years', 'income_type']].head(20)

days_employed  days_employed_years  dob_years income_type
0    -8437.673028           -23.116912         42   сотрудник
1    -4024.803754           -11.026860         36   сотрудник
2    -5623.422610           -15.406637         33   сотрудник
3    -4124.747207           -11.300677         32   сотрудник
4   340266.072047           932.235814         53   пенсионер
5     -926.185831            -2.537495         27   компаньон
6    -2879.202052            -7.888225         43   компаньон
7     -152.779569            -0.418574         50   сотрудник
8    -6929.865299           -18.985932         35   сотрудник
9    -2188.756445            -5.996593         41   сотрудник
10   -4171.483647           -11.428722         36   компаньон
11    -792.701887            -2.171786         40   сотрудник
12            NaN                  NaN         65   пенсионер
13   -1846.641941            -5.059293         54   сотрудник
14   -1844.956182            -5.054674         56   компаньон
15    -972.364419            -2.664012         26   сотрудник
16   -1719.934226            -4.712149         35   сотрудник
17   -2369.999720            -6.493150         33   сотрудник
18  400281.136913          1096.660649         53   пенсионер
19  -10038.818549           -27.503612         48   сотрудник

Значения в столбце `days_employed_years` выглядят адекватно для возраста и занятости клиента, если брать их по модулю. Ошибка со знаком могла возникнуть в результате расчетов, если в формуле из дня начала работы вычитали текщий день (т.е. из меньшего большее).

Интереснее дело обстоит с положительными значениями (аномалиями). Они могли возникнуть в результате разного форматирования дат, если данные не привели к общему формату. Или по какой-то причине быть завышены в 100 раз. В любом случае, установить истинную причину без исходных данных не удастся.

Данный столбец не будет использоваться для проверки гипотез и не требуется для дальнейших расчетов. Заниматься его преобработкой нецелесообразно. Он может быть просто удален без потери важных данных. Однако поставлена задача привести его значения к нормальному виду. Поэтому мы можем поступить с данными в столбце следующим образом:
1. Заменить аномалии на медиальные значения.
2. Заменить значения в столбце на их модули.
2. Заменить пропуски на медиальные значения.

In [142]:
# сначала удалим лишний столбец, использованный для обзора данных
df = df.drop('days_employed_years', axis=1)

In [143]:
# расчитываем медиальное значение
days_employed_median = df.loc[df['days_employed'] < 0, 'days_employed'].median()

# заменяем аномалии медиальными значениями
df.loc[df['days_employed'] > 0, 'days_employed'] = days_employed_median

# заменяем пропуски на медиальные значения
df['days_employed'] = df['days_employed'].fillna(days_employed_median)

# заменяем отрицательные значения на положительные
df['days_employed'] = abs(df['days_employed'])

In [144]:
# проверяем результат замены
print(df.info())
df.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
None


children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0    1630.019381         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Данные в столбце `days_employed` успешно преобразованы.

In [145]:
df['dob_years'].value_counts().tail(15)

66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

В столбце `dob_years` присутствует 101 строка со значением 0. Возраст клиента не может быть равен 0. Скорее всего, отсутсвующие данные в этом столбце были заменены на 0.

Данный столбец не будет использоваться для проверки гипотез и не требуется для дальнейших расчетов. Заниматься его преобработкой нецелесообразно, поэтому оставим его в исходном виде.

## Шаг 2.3. Изменение типов данных.

Для удобства и лаконичности заменим тип данных в столбцах `days_employed` и `total_income` на int.

In [146]:
# замена типа данных
df['days_employed'] = df['days_employed'].astype(int)
df['total_income'] = df['total_income'].astype(int)

# проверка результата
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int32 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int32 
 11  purpose           21525 non-null  object
dtypes: int32(2), int64(5), object(5)
memory usage: 1.8+ MB


Успешно.

## Шаг 2.4. Удаление дубликатов.

Проверим таблицу на наличие явных дубликатов. 

In [147]:
# посчет количества явных дубликатов в таблице
df.duplicated().sum()

54

Удалим их.

In [148]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index(drop=True)

Убедимся, что явных дубликтов не осталось.

In [149]:
df.duplicated().sum()

0

Теперь необходимо устранить неявные дубликаты. Проверим все столбцы на наличие неявных дублекатов.

In [150]:
# создадим список названий колонок
df_columns = df.columns

# создадим цикл для проверки значений
for column in df_columns:
    print(f'Список значений в столбце {column}:')
    print(df[column].value_counts())

Список значений в столбце children:
 0     14107
 1      4809
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64
Список значений в столбце days_employed:
1630    5570
133       16
327       16
438       15
204       14
        ... 
7076       1
5029       1
5045       1
2998       1
6235       1
Name: days_employed, Length: 5716, dtype: int64
Список значений в столбце dob_years:
35    616
40    607
41    606
34    601
38    597
42    596
33    581
39    572
31    559
36    554
44    545
29    544
30    538
48    537
37    536
50    513
43    512
32    509
49    508
28    503
45    497
27    493
52    484
56    484
47    477
54    476
46    473
53    459
57    456
58    456
51    448
59    443
55    443
26    408
60    374
25    357
61    354
62    349
63    269
24    264
64    262
23    253
65    194
22    183
66    182
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74

* В столбце `education` присутсвуют одни и те же значения в разном регистре.
* В столбце `purpose` присутсвуют большое количество смысловых дубликатов, которые могли возникнуть в результате заполлнения анкеты клиентами.
* В столбце `children` присутсвуют отрицательные значение, а также подозрительно большое количество значений 20.
* В столбце `gender` присутсвует небинарное значение.

Сначала разберемся со столбцом столбце `education`. Приведем значения к одному регистру.

In [151]:
# приведем значения в столбце 'education' к одному регистру и заменим в таблице
df['education'] = df['education'].str.lower()

# проверим результат
df.head(10)

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0           1630         53   среднее             1   
5         0            926         27    высшее             0   
6         0           2879         43    высшее             0   
7         0            152         50   среднее             1   
8         2           6929         35    высшее             0   
9         0           2188         41   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Теперь столбец `purpose`.

Для получения правильных результатов, от смысловых дубликатов необходимо избавиться, приведя однотипные значения к общему знаминателю.

In [152]:
# приведем к общему знаменателю цели на свадьбу
df.loc[df['purpose'].str.contains('свадьб'), 'purpose'] = 'свадьба'

# приведем к общему знаменателю цели, связанные с автомобилями
df.loc[df['purpose'].str.contains('авто'), 'purpose'] = 'автомобиль'

# приведем к общему знаменателю цели, связанные с недвижимостью
df.loc[df['purpose'].str.contains('недвиж'), 'purpose'] = 'операции с недвижимостью'
df.loc[df['purpose'].str.contains('жиль'), 'purpose'] = 'операции с недвижимостью'

# приведем к общему знаменателю цели, связанные с образованием
df.loc[df['purpose'].str.contains('образован'), 'purpose'] = 'образование'

Проверим результат преобразований.

In [153]:
df['purpose'].value_counts()

операции с недвижимостью    10814
автомобиль                   4308
образование                  4014
свадьба                      2335
Name: purpose, dtype: int64

Успешно.

В столбце `children` присутсвуют отрицательное значение. Количество детей не может быть отрицательным. Скорее всего это опечатка и вместо -1 должно быть 1. Количество значений с ошибкой в данном случае составляет около 1% от числа клиентов с 1 ребеноком и замена значения -1 на 1 не повлияет на результат.

In [154]:
# заменим значение -1 на 1 в столбце children
df.loc[df['children'] == -1, 'children'] = 1

Также в столбце присутствует 76 строк со значением 20, что является пятым по встречаимости значением. Учитывая, что другие значения находятся в интервале от 0 до 5, резкий скачок к 20 выглядит нериалистично. Вероятняее всего закралась ошибка, например, при форматировании - стояло значение типа float 2.0, которое превратилось в 20. Заменим 20 на 2.

In [155]:
# заменим значение 20 на 2 в столбце children
df.loc[df['children'] == 20, 'children'] = 2

In [156]:
df['children'].value_counts()

0    14107
1     4856
2     2128
3      330
4       41
5        9
Name: children, dtype: int64

Успешно.

В столбце `gender` присутсвует одно небинарное значение. Попробуем установить его истиное значение по другим столбцам.

In [157]:
# выведем строку со значением XNA в столбце gender
df[df['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10690         0           2358         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10690  гражданский брак                 1    XNA   компаньон     0   

       total_income                   purpose  
10690        203905  операции с недвижимостью

In [158]:
# проверим, есть ли связь между гендером и типом занятости
df[df['income_type'] == 'компаньон'].head(5)

children  days_employed  dob_years education  education_id  \
5          0            926         27    высшее             0   
6          0           2879         43    высшее             0   
10         2           4171         36    высшее             0   
14         0           1844         56    высшее             0   
33         0           1548         48   среднее             1   

       family_status  family_status_id gender income_type  debt  total_income  \
5   гражданский брак                 1      M   компаньон     0        255763   
6    женат / замужем                 0      F   компаньон     0        240525   
10   женат / замужем                 0      M   компаньон     0        113943   
14  гражданский брак                 1      F   компаньон     1        165127   
33   женат / замужем                 0      F   компаньон     0        157245   

                     purpose  
5   операции с недвижимостью  
6   операции с недвижимостью  
10  операции с недвижимостью  
14  операции с недвижимостью  
33  операции с недвижимостью

Проверка не дала результата.

В таком случае имеет смысл заменить данное значение, на F, так как из в столбце в 2 раза больше чем M. Статистически это не скажется на результатах, тем более, что данный столбец не используется в расчетах.

In [159]:
# заменим значение XNA на F
df.loc[df['gender'] == 'XNA', 'gender'] = 'F'

# и проверим результат
df.loc[10690]

children                                   0
days_employed                           2358
dob_years                                 24
education                неоконченное высшее
education_id                               2
family_status               гражданский брак
family_status_id                           1
gender                                     F
income_type                        компаньон
debt                                       0
total_income                          203905
purpose             операции с недвижимостью
Name: 10690, dtype: object

Успешно.

Повторно прверим всю таблицу на предмет явных дубликатов, т.к. они могли появиться, после исправления неявных дубликатов.

In [160]:
df.duplicated().sum()

351

Появились новые дубликаты. Избавимся от них.

In [161]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index(drop=True)

Убедимся, что явных дубликтов не осталось.

In [162]:
df.duplicated().sum()

0

Получим информацию об обновленной таблице перед тем, как двигаться дальше.

In [163]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21120 entries, 0 to 21119
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21120 non-null  int64 
 1   days_employed     21120 non-null  int32 
 2   dob_years         21120 non-null  int64 
 3   education         21120 non-null  object
 4   education_id      21120 non-null  int64 
 5   family_status     21120 non-null  object
 6   family_status_id  21120 non-null  int64 
 7   gender            21120 non-null  object
 8   income_type       21120 non-null  object
 9   debt              21120 non-null  int64 
 10  total_income      21120 non-null  int32 
 11  purpose           21120 non-null  object
dtypes: int32(2), int64(5), object(5)
memory usage: 1.8+ MB


Количество строк уменьшилось, пропусков нет, типы данных изменились.

## Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

На данный момент идентификаторы и расшифровки для уровня образования и семейного статуса хранятся в одной таблице.

Создадим "словари" - два новых датафрейма, в которых:
* в первом: каждому уникальному значению из `education` будет соответствовать уникальное значение `education_id`;
* во втором: каждому уникальному значению из `family_status` будет соответствовать уникальное значение `family_status_id`.

Затем удалим столбцы `education` и `family_status`, оставив в таблице только идентификаторы.

In [164]:
# создадим словарь для идентификации уровня образования клиента
education_dict = df[['education_id', 'education']]

# удалим дубликаты из словаря (с удалением старых индексов и формированием новых)
education_dict = education_dict.drop_duplicates().reset_index(drop=True)

# создадим словарь для идентификации семейного положения клиента
family_status_dict = df[['family_status_id', 'family_status']]

# удалим дубликаты из словаря (с удалением старых индексов и формированием новых)
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)

Убедимся, что словари сформированы верно.

In [165]:
display(education_dict)
display(family_status_dict)

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

Удалим столбцы с рашифровкой идентификаторов из общей таблицы.

In [166]:
df = df.drop(columns=['education', 'family_status'], axis=1)

In [167]:
# убедимся, что столбцы удалены
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           1630         53             1                 1      F   

  income_type  debt  total_income                   purpose  
0   сотрудник     0        253875  операции с недвижимостью  
1   сотрудник     0        112080                автомобиль  
2   сотрудник     0        145885  операции с недвижимостью  
3   сотрудник     0        267628               образование  
4   пенсионер     0        158616                   свадьба

Успешно.

## Шаг 2.6. Категоризация дохода.

На основании диапазонов, предоставленных заказчиком, создадим столбец `total_income_category` с категориями по уровню дохода:
* 0–30000 — `E`;
* 30001–50000 — `D`;
* 50001–200000 — `C`;
* 200001–1000000 — `B`;
* 1000001 и выше — `A`.

In [119]:
# создадим функцию для категоризации
def total_income_categorisation(income):
    try:
        if 0 <= income < 30001:
            return 'E'
        if income < 50001:
            return 'D'
        if income < 200001:
            return 'C'
        if income < 1000001:
            return 'B'
        if income > 1000000:
            return 'A'
    except:
        print('Отрицательный уровень дохода!')

# применим функцию к таблице
df['total_income_category'] = df['total_income'].apply(total_income_categorisation)

Проверим результат в таблице.

In [120]:
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0           1630         53             1                 1      F   

  income_type  debt  total_income                   purpose  \
0   сотрудник     0        253875  операции с недвижимостью   
1   сотрудник     0        112080                автомобиль   
2   сотрудник     0        145885  операции с недвижимостью   
3   сотрудник     0        267628               образование   
4   пенсионер     0        158616                   свадьба   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

Успешно.

## Ответы на вопросы.

##### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

Чтобы оценить влияние количества детей клиента на наличие задолжностей, проведем анализ по столбцам `children` и `debt`.

Для этого сгрупперуем данные по столбцу `children` и расчитаем для каждой группы общее число клиентов и суммарное число должников по столбцу `debt`.

In [121]:
debt_to_children = df.groupby('children').agg({'debt': ['sum', 'count']})
debt_to_children

debt       
           sum  count
children             
0         1061  13829
1          445   4798
2          202   2115
3           27    329
4            4     40
5            0      9

Расчитаем число должников в процентном соотношении для каждой группы.

In [122]:
# debt_to_children_persent - процент должников среди клиентов с количеством детей, обозначенным цифрой в левом столбце
debt_to_children_persent = debt_to_children['debt']['sum'] / debt_to_children['debt']['count'] * 100
debt_to_children_persent.sort_values(ascending=False)

children
4    10.000000
2     9.550827
1     9.274698
3     8.206687
0     7.672283
5     0.000000
dtype: float64

##### Вывод 1:

По полученным данным видно, что наибольшее количество должников среди семей с 4 детьми - 10%.

Далее следуют 2 и 1 ребенок - 9.55, 9.27 % соответсвенно.

Следующие 3 и 0 детей - 8.21, 7.67 % соответственно.

Среди людей с 5 детьми должников не оказалось. Однако таких семей в выборке всего 9. Утверждать, что данные достоверны - нельзя. Возможен статистический выброс.

Из полученных данных можно сделать вывод, что клиенты без детей возвращают кредиты в срок чаще, чем клиенты с 4 детьми.
Тенденция к увеличению числа просрочек с возрастанием количества детей прослеживается не совсем четко. Наиболее кредитоспособными по текущим данным выглядят клиенты у которых 0, 3 или 5 детей.

##### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок:

Чтобы оценить зависимость наличия задолжностей от семейного положения, необходимо провести анализ по столбцам `family_status` и `debt`. Сначала расшифровки значений `family_status_id` необходимо извлечь из словаря `family_status_dict`.

Затем в полученной таблице сгрупперуем данные по столбцу `family_status` и расчитаем для каждой группы общее число клиентов и суммарное число должников по столбцу `debt`.

In [123]:
# создадим таблицу со столбцами family_status_id и debt
debt_to_family_status_id = df[['family_status_id', 'debt']]

# добавим в таблицу расшифровки столбца family_status_id по словарю family_status_dict
debt_to_family_status = debt_to_family_status_id.merge(family_status_dict, on='family_status_id', how='left')

# удалим столбец family_status_id
debt_to_family_status = debt_to_family_status.drop('family_status_id', axis=1)

# сгруппируем данные по столбцу `family_status` и расчитаем для каждой группы общее число клиентов и суммарное число должников по столбцу `debt`
debt_to_family_status = debt_to_family_status.groupby('family_status').agg({'debt': ['sum', 'count']})

debt_to_family_status

debt       
                       sum  count
family_status                    
Не женат / не замужем  274   2784
в разводе               85   1193
вдовец / вдова          63    943
гражданский брак       388   4124
женат / замужем        929  12076

Расчитаем число должников в процентном соотношении для каждой группы.

In [124]:
# debt_to_family_status_persent - процент должников среди клиентов в соответствии с их семейным положением
debt_to_family_status_persent = debt_to_family_status['debt']['sum'] / debt_to_family_status['debt']['count'] * 100
debt_to_family_status_persent.sort_values(ascending=False)

family_status
Не женат / не замужем    9.841954
гражданский брак         9.408341
женат / замужем          7.692945
в разводе                7.124895
вдовец / вдова           6.680806
dtype: float64

##### Вывод 2:

По полученным данным видно, что наибольшее количество должников среди клиентов не состоящих в браке или проживающих в гражданском браке - 9.84 и 9.41 % соответсвенно.

Среди клиентов состоящих в официальном браке и разведенных процент просрочек по заемам примерно на 2 пункта ниже - 7.69 и 7.12 % соответсвенно.

Самое низкое число просрочек у вдовцов и вдов - 6.68%.

Из полученных данных можно сделать вывод, число просрочек по выплате займов коррелирует с семейным положением клиентов.

##### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок:

Чтобы оценить зависимость наличия задолжностей от уровня дохода, проведем анализ по столбцам `total_income_category` и `debt`.

Для этого сгрупперуем данные по столбцу `total_income_category` и расчитаем для каждой группы общее число клиентов и суммарное число должников по столбцу `debt`.

In [125]:
debt_to_total_income_category = df.groupby('total_income_category').agg({'debt': ['sum', 'count']})
debt_to_total_income_category

debt       
                        sum  count
total_income_category             
A                         2     25
B                       356   5042
C                      1358  15681
D                        21    350
E                         2     22

Расчитаем число должников в процентном соотношении для каждой группы.

In [126]:
# debt_to_total_income_category_persent - процент должников среди клиентов в соответствии с их уровнем дохода
debt_to_total_income_category_persent = debt_to_total_income_category['debt']['sum'] / debt_to_total_income_category['debt']['count'] * 100
debt_to_total_income_category_persent.sort_values(ascending=False)

total_income_category
E    9.090909
C    8.660162
A    8.000000
B    7.060690
D    6.000000
dtype: float64

##### Вывод 3:

По полученным данным видно, что наибольшее количество должников среди клиентов с категорией доходов `Е` (от 0 до 30 000) - 9.09%.

Далее следуют категории `С`(от 50 001 до 200 000) и `А`(свыше 1 000 000) - 8.66, 8% соответсвенно.

Ещё ниже у `В`(от 200 001 до 1 000 000) - 7.06%.

Самый низкий процент прсрочек у клиентов с категорией доходов `D`(от 30 001 до 50 000) - 6%.


Из полученных данных можно сделать вывод, существует нелинейная зависимость между уровнем дохода и просрочкой по выплатам. Так, наименьшее число просрочек у клиентов с доходами от 30 001 до 50 000 и от от 200 001 до 1 000 000 (категории `D` и `B`).

Несколько выше у клиентов с доходами от от 50 001 до 200 000 и свыше 1 000 000 (категории `C` и `A`).

Самое высокое число просрочек у клиентов с минимальным доходом до 30 000 рублей (категория `E`).

При этом большинство клиентов по уровню дохода принадлежат категориям `B` и `C`.

##### Вопрос 4: Как разные цели кредита влияют на его возврат в срок:

Чтобы оценить зависимость наличия задолжностей от цели кредита, проведем анализ по столбцам `purpose` и `debt`.

Для этого сгрупперуем данные по столбцу `purpose` и расчитаем для каждой группы общее число клиентов и суммарное число должников по столбцу `debt`.

In [127]:
debt_to_purpose = df.groupby('purpose').agg({'debt': ['sum', 'count']})
debt_to_purpose

debt       
                          sum  count
purpose                             
автомобиль                402   4272
образование               370   3964
операции с недвижимостью  781  10578
свадьба                   186   2306

Расчитаем число должников в процентном соотношении для каждой группы.

In [128]:
# debt_to_purpose_persent - процент задолжностей в зависимости от цели кредита
debt_to_purpose_persent = debt_to_purpose['debt']['sum'] / debt_to_purpose['debt']['count'] * 100
debt_to_purpose_persent.sort_values(ascending=False)

purpose
автомобиль                  9.410112
образование                 9.334006
свадьба                     8.065915
операции с недвижимостью    7.383248
dtype: float64

##### Вывод 4:

Согласно полученным данным существует зависимость между целью кредита и риском возникновения задолжности.

Так наибольшее число просрочек приходится на кредитования расходов на автомобили и образование - 9.41 и 9.33% соответсвенно.

Ниже число задолжносей при заеме на свадебные расходы - 8.07%.

Самое низкое для заемов на операции с недвижимостью - 7.38%.

## Общий вывод:

Мы проанализировали 4 гипотезы и выявили следующее:
1. Данные по количеству детей плохо коррелируют с возвратом кредита в срок. Тем не менее, установлено, что меньше всего задолжностей у клиентов без детей (если опустить данные по 9 клиентм с 5 детьми), а больше всего у клиентов с 4 детьми. Однако внутри групп от 1 до 5 детей прямая зависимость отсутсвует.<br>Не рекомендуется ориентироваться на количесво детей при составлении системы оценки.


2. Зависимость между семейным положением и возвратом кредита в срок присутствует и ярко выражена. Так минимальное число должников среди вдов и вдовцов, максимальное - среди холостых и проживающих в гражданском браке.


3. Зависимость между уровнем дохода и возвратом кредита в срок существует, но не является линейной. Так максимальное число просрочек у клиентов с самым низким доходом категория `E`, а минимальное - у следующей в порядке увеличения доходов категории `D`.


4. Установлено, что цели кредита действительно влияют на его возврат в срок. Наибольшее число задолжностей по кредитам на операции с автомобилями, а наименьшее - с недвижимостью.
